### PR TITLE
[None][fix] Structured Output with DSpark Speculative Drafter

### LINKED ISSUE #16967: [Bug]: Structured Output with DSpark Speculative Drafter
### System Info

v1.3.0rc22 release container

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

Launch trtllm with (replace with your weight dir) :
```
docker run -d --name dspark --gpus '"device=0"' --ipc=host --ulimit memlock=-1 --shm-size=16g -v /data:/data -p 8000:8000 nvcr.io/nvidia/tensorrt-llm/release:1.3.0rc22 trtllm-serve serve <weight_dir> --served_model_name=deepseek-ai/DeepSeek-V4-Flash --backend=pytorch --tp_size=1 --ep_size=1 --max_seq_len=163840 --max_num_tokens=16384 --max_batch_size=72 --enable_chunked_prefill --trust_remote_code --custom_tokenizer=deepseek_v4 --host=0.0.0.0 --port=8000 --extra_llm_api_options dspark.json
```

with `dspark.json`:
```
{"print_iter_log":true,"kv_cache_config":{"tokens_per_block":128,"host_cache_size":0,"dtype":"fp8","free_gpu_memory_fraction":0.85,"enable_block_reuse":true,"enable_partial_reuse":false},"sampler_type":"auto","scheduler_config":{"capacity_scheduler_policy":"MAX_UTILIZATION","waiting_queue_policy":"priority"},"speculative_config":{"decoding_type":"DSpark","max_draft_len":5,"speculative_model":"<weight_dir>"},"cuda_graph_config":{"enable_padding":true},"moe_config":{"backend":"TRTLLM"},"guided_decoding_backend":"xgrammar","disable_overlap_scheduler":true}
```

After the engine launch, request the engine with:

```
curl -s -X POST http://localhost:8000/v1/chat/completions -H 'Content-Type: application/json' -d '{"model":"deepseek-ai/DeepSeek-V4-Flash","messages":[{"role":"user","content":"Describe machine learning in one paragraph, then rate its difficulty."}],"response_format":{"type":"json_schema","json_schema":{"name":"desc","strict":true,"schema":{"type":"object","additionalProperties":false,"properties":{"topic":{"type":"string"},"level":{"type":"string","enum":["beginner","intermediate","advanced"]},"score":{"type":"integer"}},"required":["topic","level","score"]}}},"max_tokens":160,"temperature":0.7}'
```

Response (this can be stochastic since the 0.7 temperature):
```
{"id":"chatcmpl-57439ca490474f97a0f3f366217c28fc","object":"chat.completion","created":1785260834,"model":"deepseek-ai/DeepSeek-V4-Flash","choices":[{"index":0,"message":{"role":"assistant","content":"{\n  \"topic\": \"Machine learning\",\n  \"level\": \"intermediate\"\n}\n\nMachine learning is a subset of artificial intelligence that enables computers to learn patterns and make decisions from data without being explicitly programmed for every task. Instead of following rigid, hand-coded rules, algorithms are trained on large datasets, adjusting their internal parameters to minimize errors in predictions or classifications. This process allows systems to improve over time, powering applications like spam filters, recommendation engines, and self-driving cars.\n\n**Difficulty rating:** Intermediate. Understanding the core concept is straightforward, but mastering the underlying mathematics (linear algebra, calculus, probability), algorithm selection, and practical implementation requires significant study and experience.","reasoning_content":"","reasoning":null,"tool_calls":[]},"logprobs":null,"finish_reason":"stop","stop_reason":null,"mm_embedding_handle":null,"disaggregated_params":null,"avg_decoded_tokens_per_iter":2.914893627166748}],"usage":{"prompt_tokens":16,"total_tokens":153,"completion_tokens":137,"prompt_tokens_details":{"cached_tokens":0}},"prompt_token_ids":null,"prompt_token_ids_b64":null}
```

### Expected behavior

Following the schema indicated in the request.

### actual behavior

In the above example, the score field in the response is not present and violating the schema.

### additional notes

N/A

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by chungen04**
### Description

Ref #15808  
Fix #16967

This PR fixes DSpark speculative decoding structured-output guided decoding by applying xgrammar guidance consistently during draft-token generation, so grammar constraints remain enforced across the full speculative block (and JSON-schema violations like missing required fields are avoided).

Previously, DSpark applied the xgrammar-derived mask only when performing target verification sampling, while the drafter sampled draft tokens without grammar constraints. If a drafted token became grammar-illegal, the matcher/bitmask state could become inconsistent within the block, allowing later verification to proceed without the intended constraints and potentially committing an illegal token (silently disabling guidance).

Now, when `guided_decoder` is present, DSpark generates draft tokens **left-to-right** for each of the `K` draft steps:
- seeds matcher state from each request’s last accepted token
- for each draft position `k`:
  - calls `guided_decoder.add_draft_batch(..., draft_step=k)`
  - calls `guided_decoder.execute_draft_batch(...)` to apply the per-position grammar mask **in-place**
  - samples the drafter token from the masked, guided step logits
- returns only generation-row tokens (excluding context rows)

### Key implementation changes

- Added `_sample_draft_tokens_guided(...)` in `tensorrt_llm/_torch/speculative/dspark.py`:
  - converts provided `gen_logits` into a step-major layout to produce per-step `[batch, vocab]` tensors (with zero-filled context rows when `num_contexts > 0`)
  - seeds matcher using `accepted_tokens` last-accepted token per request
  - iterates `k in 0..K-1`: `add_draft_batch` → `execute_draft_batch` → `sample_draft_tokens`
  - assembles and returns `gen_draft_tokens` as `[num_gens, K]`
- Updated `DSparkWorker._forward_impl(...)` to route draft sampling:
  - guided path: `_sample_draft_tokens_guided(...)`
  - fallback path: existing unguided `sample_draft_tokens(...)`
- Preserved the `gen_vocab`/`draft_probs_last_dim` handling and its tensor-parallel comment to ensure the context one-hot uses the **post-TP scatter width** (`spec_metadata.draft_probs_last_dim`), not the sharded `gen_logits` width.

## Dev Engineer Review

- **Correctness:** Implements guided, per-draft-step left-to-right sampling for DSpark so `guided_decoder` updates/masks the logits *before* the drafter samples at each step, keeping matcher/verification state synchronized across the entire `K` block (directly addressing the guidance-desynchronization failure mode behind JSON-schema violations).
- **Consistency:** Guided decoding now mirrors the guided per-step loop ordering (`add_draft_batch` → `execute_draft_batch` → sampling), reducing the risk of divergence between drafter drafts and matcher/verify logic.
- **TP/shape correctness:** Builds step-major logits and preserves the existing logic around `gen_vocab = spec_metadata.draft_probs_last_dim`, including the comment about matching the context one-hot scatter width under tensor parallelism.
- **Test coverage gaps / risk areas noted in review request:** The new unit test covers mixed context+generation guided drafting behavior, but explicit coverage for “matcher rollback on the final step” and guided decoding with “tensor-parallel rejection-probability storage under CUDA-graph capture” is not confirmed in the added test.

### Test changes
- **Added:** `test_forward_guided_batch_masks_and_advances_matcher_per_step` in `tests/unittest/_torch/speculative/hw_agnostic/test_dspark_worker.py`
  - Validates guided mixed (context + gen) batching:
    - left-to-right `K` draft step ordering
    - `add_draft_batch` called per draft position
    - `execute_draft_batch` invoked with contiguous `[batch, vocab]` logits per step and zero-summed context rows
    - matcher seeding uses each request’s last accepted token
    - in-place grammar masking occurs **before** sampling (forbidden token set to `-inf`)
    - context one-hot drafting uses scatter “full vocab” width (`draft_probs_last_dim`)
    - assembled `next_draft_tokens` excludes context rows and never proposes the forbidden token

### Coverage in tests/integration/test_lists/
- This specific unittest is **not** referenced in `tests/integration/test_lists/test-db/` or `tests/integration/test_lists/qa/` (only DSpark appears in a performance QA list unrelated to this unit test).

**Verdict:** needs follow-up.

### CODE REVIEW

**Comment by chungen04** [comment]
Will rebase to main

**Comment by BowenFu** [comment]
The diagnosis is right and the fix is at the right layer — masking only the target-verify logits while the drafter proposes unconstrained tokens does break the matcher mid-block and silently disable guidance from there on. Walking the K positions left-to-right with `add_draft_batch`/`execute_draft_batch` before each sample is the same contract MTP and Eagle3 already use.

I went through the new path and did not find a correctness defect: `_draft_gen_block_batched` returns `[num_gens, K, vocab]`, so the transpose lines up with `full_logits[:, num_contexts:, :]`; the `num_contexts == 0` branch is equivalent since `num_gens == batch_size` there; passing a 2-D per-step tensor selects the same sampler path MTP uses, so `maybe_gather_sharded_draft_logits` still gathers before the scatter into `draft_probs[..., draft_step, :]`; and the context rows are transient (excluded from `gen_draft_tokens`, overwritten by `write_context_onehot_draft_probs`, and rolled back by `execute_draft_batch` on the last step). The unguided `else:` branch is unchanged apart from the shortened comment.

What's missing is a test. Test Coverage is blank, and the existing DSpark tests explicitly set `worker.guided_decoder = None`, so nothing exercises this. The subtle parts are exactly the ones static reading can't settle: the mixed context/generation batch, the matcher rollback on the final step, and guided + TP rejection-probability storage under CUDA-graph capture. A focused unit test over `_sample_draft_tokens_guided` with `num_contexts > 0` would cover most of it.

Worth keeping the deleted comment on the `else:` branch, too — the reason `gen_vocab` must be `draft_probs_last_dim` and not `gen_logits.shape[-1]` under TP is not obvious, and it now applies to both branches.

**Comment by chungen04** [comment]
@BowenFu Thank you for the suggestions. Addressing:
> Worth keeping the deleted comment on the else: branch, too — the reason gen_vocab must be draft_probs_last_dim and not gen_logits.shape[-1] under TP is not obvious, and it now applies to both branches.
- The change is slightly refactored. The comment was restored so it is unchanged since orignial, and the assignment `gen_vocab = spec_metadata.draft_probs_last_dim` was moved out from `if`/`else` since it was independent of whether the decoding is guided.
> What's missing is a test. Test Coverage is blank, and the existing DSpark tests explicitly set worker.guided_decoder = None, so nothing exercises this. The subtle parts are exactly the ones static reading can't settle: the mixed context/generation batch, the matcher rollback on the final step, and guided + TP rejection-probability storage under CUDA-graph capture. A focused unit test over _sample_draft_tokens_guided with num_contexts > 0 would cover most of it.
- Added unit test. Additionally, I would like to raise: (1) e2e test for DSpark does not cover guided decoding, only an unittest is shipped. See [test_llm_api_pytorch.py](https://github.com/NVIDIA/TensorRT-LLM/blob/c45ad837c5f9a705f1db6149e7281544f4cd8835/tests/integration/defs/accuracy/test_llm_api_pytorch.py#L3962). (2) for MTP, unit test is also unavailable in [test_mtp.py](https://github.com/NVIDIA/TensorRT-LLM/blob/c45ad837c5f9a705f1db6149e7281544f4cd8835/tests/unittest/_torch/speculative/hw_agnostic/test_mtp.py). Given that these tests are out of scope in this PR, I will defer it to future work.

**Comment by BowenFu** [comment]
Thanks for adding `test_forward_guided_batch_masks_and_advances_matcher_per_step` — that's exactly the coverage I was asking for, and it pins the parts static reading can't settle (mask-before-sample ordering, the contiguous `[batch, vocab]` slice, zero-padded context rows, the matcher seed, and the `draft_probs_last_dim` scatter width).

One blocker before this lands, in the new helper:

```python
gidx = (num_accepted_tokens - 1).clamp(min=0).unsqueeze(1)
new_tokens = accepted_tokens.gather(1, gidx).squeeze(1).to(torch.int32)
```

`num_accepted_tokens` is allocated `dtype=torch.int` (int32) in `SpecWorkerBase.sample_and_accept_draft_tokens` (`interface.py:1365-1367`), so `gidx` is int32 — but `Tensor.gather` requires an int64 index. The guided path raises before it drafts anything.

The existing DSpark code does the conversion explicitly for the identical operation a few hundred lines up:

```python
nacc = num_accepted_tokens[num_contexts:batch_size].long()   # dspark.py:406
gidx = nacc - 1
bonus = accepted_tokens[...].gather(1, gidx.unsqueeze(1)).squeeze(1).long()
```

So it's a one-line fix — `.long()` on `gidx` (or on `num_accepted_tokens`) before the gather.

Worth flagging that your own new test would have caught this: it supplies `num_accepted = torch.ones(batch_size, dtype=torch.int32)`, and `unittest/_torch/speculative/hw_agnostic` is in `l0_h100.yml:53`. But the head that adds the test (`4c31cab9e`) has never had a pipeline run on it — all three `/bot run`s were on `5cf8ba2`, and the only checks on the current head are lint/title/DCO. Please trigger a run after the fix rather than relying on the earlier green.

Everything else still looks right to me: the default path is untouched (`guided_decoder` is `None` from `SpecWorkerBase.__init__` and only set via `set_guided_decoder`, so `_forward_impl` takes the `else` branch with byte-identical arguments), the step-major layout math checks out, and keeping the `draft_probs_last_dim` comment where it covers both branches reads well.

**Comment by chungen04** [comment]
@BowenFu Fixed, thanks for the catch. Looks like the runtime in torch 2.11.0 takes int32 index in `gather`, but per [docs](https://docs.pytorch.org/docs/2.11/generated/torch.gather.html) I am still adding the cast, while matching the existing lines (`# dspark.py:406` as you mentioned).

**Review by BowenFu — APPROVED** [review]

**Comment by chungen04** [comment]
@BowenFu can you also trigger the CI? Thanks.

**Comment by BowenFu** [comment]
The `.long()` cast on `95d3f17b5` clears my blocker — `gidx` is now int64 for the `gather`, and the `.to(torch.int32)` on the result keeps `new_tokens` in the dtype `add_draft_batch` gets everywhere else, so it matches `dspark.py:406` exactly. Nothing else on the new guided path indexes with an int32 tensor.

One thing to be aware of before this merges: **there is no `blossom-ci` run on `95d3f17b5`, and there wasn't one on `4c31cab9e` either.** The three green `/bot run`s referenced earlier all ran on `5cf8ba265` — i.e. on the commit *before* the new test and before the gather fix. So the new test has not actually executed in CI yet. Worth a full run on the current head rather than relying on the earlier greens.

Remaining non-blocking note, unchanged from before: the fake decoder in the test doesn't model the real matcher's rollback at its final draft step (`guided_decoder.py:583-603`), so the test can't prove the per-step `add_draft_batch`/`execute_draft_batch` loop leaves the matcher in the state the subsequent verify expects. Static reading says it does; I'm flagging it as the one thing the coverage can't settle, not asking for more tests.

File: tensorrt_llm/_torch/speculative/dspark.py [hunk 1]
```diff
@@ -453,6 +404,44 @@ def _draft_gen_block_batched(
         )
         return block_logits
 
+    def _sample_draft_tokens_guided(
```
Comment (cascade812) [inline tensorrt_llm/_torch/speculative/dspark.py:407]: Has this exact path been tested with cuda_graph_config.enable_padding=true, XGrammar, and non-greedy sampling? The new loop performs K capturable guided-decoder callbacks inside DSpark’s previously block-oriented path.

**Review by cascade812 — APPROVED** [review]

File: tensorrt_llm/_torch/speculative/dspark.py (hunk 1 above)
Comment (chungen04) [reply tensorrt_llm/_torch/speculative/dspark.py:407]: Verified empirically, ran DSpark with guided_decoding_backend=xgrammar, cuda_graph_config.enable_padding=true, and non-greedy sampling (temperature=0.8, top_p=0.95), then fired 20 concurrent JSON-schema requests

- 20/20 outputs valid and schema-conforming
- avg_decoded_tokens_per_iter ≈ 2.8–3.8 across all requests

So the K per-step guided callbacks capture and replay correctly under enable_padding. This was a single-GPU run with the TRTLLM MoE backend.

**Comment by chungen04** [comment]
Please trigger the CI -- thanks! cc @bo-nv @BowenFu @cascade812 @brnguyen2 .

**Comment by chungen04** [comment]
Hi, is it ready to merge?
